igor-web/src/main/groovy/com/netflix/spinnaker/igor/wercker/WerckerCache.java

```diff
@@ -0,0 +1,210 @@
+     * @param appAndPipelineName
+     * @param runs
+     * @return a map containing the generated build numbers for each run created,
+     *         keyed by run id
+     */
+    public Map<String, Integer> updateBuildNumbers(String master, String appAndPipelineName, List<Run> runs) {
+        String key = makeKey(master, appAndPipelineName) + ":runs";
+        final Map<String, String> existing = redisClientDelegate.withCommandsClient(c -> {
+            if (!c.exists(key)) {
+                return null;
+            }
+            return c.hgetAll(key);
+        });
+        List<Run> newRuns = runs;
+        if (existing != null && existing.size() > 0) {
+            newRuns = runs.stream().filter(run -> !existing.containsKey(run.getId())).collect(Collectors.toList());
+        }
+        Map<String, Integer> runIdToBuildNumber = new HashMap<>();
+        int startNumber = (existing == null || existing.size() == 0) ? 0 : existing.size();
+        newRuns.sort(runStartedAtComparator);
+        for (int i = 0; i < newRuns.size(); i++) {
+            int buildNum = startNumber + i;
+            setBuildNumber(master, appAndPipelineName, newRuns.get(i).getId(), buildNum);
+            runIdToBuildNumber.put(newRuns.get(i).getId(), buildNum);
+        }
+        return runIdToBuildNumber;
+    }
+
+    public void setBuildNumber(String master, String pipeline, String runID, int number) {
+        String key = makeKey(master, pipeline) + ":runs";
+        redisClientDelegate.withCommandsClient(c -> {
+            c.hset(key, runID, Integer.toString(number));
+        });
+    }
+
+    public Long getBuildNumber(String master, String pipeline, String runID) {
+        return redisClientDelegate.withCommandsClient(c -> {
+            String ts = c.hget(makeKey(master, pipeline) + ":runs", runID);
+            return ts == null ? null : Long.parseLong(ts);
+        });
+    }
+
+    public Boolean getEventPosted(String master, String job, String runID) {
+        String key = makeEventsKey(master, job);
+        return redisClientDelegate.withCommandsClient(c -> c.hget(key, runID) != null);
+    }
+
+    public void setEventPosted(String master, String job, String runID) {
+        String key = makeEventsKey(master, job);
+        redisClientDelegate.withCommandsClient(c -> {
+            c.hset(key, runID, "POSTED");
+        });
+    }
+
+    public void pruneOldMarkers(String master, String job, Long cursor) {
+        remove(master, job);
+        redisClientDelegate.withCommandsClient(c -> {
+            c.del(makeEventsKey(master, job));
+        });
+    }
+
+    public void remove(String master, String job) {
+        redisClientDelegate.withCommandsClient(c -> {
+            c.del(makeKey(master, job));
+        });
+    }
+
+    private String makeEventsKey(String master, String job) {
+        return makeKey(master, job) + ":" + POLL_STAMP + ":events";
+    }
+
+    private String makeKey(String master, String job) {
```

Comment:
This looks to be the same key naming convention as Jenkins.  Are we sure we won't see conflicts for a user running both Jenkins and Wercker, if they have jobs with the same (or similar) names?

Reply:
Good point. Does spinnaker/halyard allows two ci masters of different types to use the same name? Anyway I added "_wercker" to prefix to avoid conflict.

Reply:
I'm actually not sure if Halyard would let you configure that, but adding the "_wercker" prefix as you've done is safer anyway.